spec.md

```diff
@@ -124,27 +124,45 @@ Typically, the first step in pulling an artifact is to retrieve the manifest. Ho
 
 ##### Pulling manifests
 
-To pull a manifest, perform a `GET` request to a url in the following form:
+To pull a manifest, perform a `GET` request to a URL in the following form:
 `/v2/<name>/manifests/<reference>` <sup>[end-3](#endpoints)</sup>
 
 `<name>` refers to the namespace of the repository. `<reference>` MUST be either (a) the digest of the manifest or (b) a tag name.
 The `<reference>` MUST NOT be in any other format.
 
 A GET request to an existing manifest URL MUST provide the expected manifest, with a response code that MUST be `200 OK`.
+A successful response SHOULD contain the digest of the uploaded blob in the header `OCI-Content-Digest`. For legacy
+reasons, the digest MAY be contained in the header `Docker-Content-Digest` instead. Either header is OPTIONAL.
 
 If the manifest is not found in the registry, the response code MUST be `404 Not Found`.
 
-##### Pulling Blobs
+##### Pulling blobs
 
-To pull a blob, perform a `GET` request to a url in the following form:
+To pull a blob, perform a `GET` request to a URL in the following form:
 `/v2/<name>/blobs/<digest>` <sup>[end-2](#endpoints)</sup>
 
 `<name>` is the namespace of the repository, and `<digest>` is the blob's digest.
 
 A GET request to an existing blob URL MUST provide the expected blob, with a response code that MUST be `200 OK`.
+A successful response SHOULD contain the digest of the uploaded blob in the header `OCI-Content-Digest`. For legacy
+reasons, the digest MAY be contained in the header `Docker-Content-Digest` instead. Either header is OPTIONAL. However,
+if one of these headers is present, the value of the header MUST be a digest matching that of the response body.
 
 If the blob is not found in the registry, the response code MUST be `404 Not Found`.
 
+##### Checking if content exists in the registry
+
+In order to verify that a repository contains a given manifest or blob, make a `HEAD` request to a URL in the following form:
+
+`/v2/<name>/manifests/<reference>` <sup>[end-12](#endpoints)</sup> (for manifests), or
+
+`/v2/<name>/blobs/<digest>` <sup>[end-11](#endpoints)</sup> (for blobs).
+
+A HEAD request to an existing blob or manifest URL MUST return `200 OK`. A successful response SHOULD contain the digest
+of the uploaded blob in the header `OCI-Content-Digest`. For legacy reasons, the digest MAY be contained in the header
+`Docker-Content-Digest` instead. Either header is OPTIONAL.
+
+If the blob or manifest is not found in the registry, the response code MUST be `404 Not Found`.
```

Comment:
As to supporting `HEAD` requests _in general_, I don't think registries `MUST` support `HEAD` requests. I would make that _optional_ (`SHOULD` or `MAY`). There's some reasons for that;

1. The `HEAD` requests are an _optimization_. Nothing should break if they're not supported, but clients may perform a pull where they could have used their local cache if they do not use `HEAD` requests, or if the registry doesn't support them.
2. `HEAD` requests may not be available in all situations (I've had users report that `HEAD` requests were blocked by proxies)
3. The spec should probably give some guidance on how clients should handle `HEAD` requests; the standard pattern is that clients _must_ (_should_ ?) fall back to `GET` requests if a `HEAD` request fails

_However_ `3.` is where a (big) challenge is; both to keep it _optional_, and to remain backward compatible with current implementations of registries.

The situation is that any current registry may not support `HEAD` requests, and because of that return a `404 Not found`. Because of that, `404` is not ambiguous, as it could either mean that;

a. the registry supports `HEAD` requests, but the given manifest was not found
b. the registry does not implement `HEAD` requests
c. `HEAD` requests are blocked (?)

This may be a problem, because now effectively clients `MUST` try the `GET` endpoint on any `404` status to know if the manifest exists (if the `GET` response also returns a `404`, the manifest does not exist).

Not sure what the best answer is  to distinguish `404 "not implemented"` from `404 "no such manifest"`. Perhaps the registry should return an empty `OCI-Manifest-Digest: ""` header to indicate _"I looked, but couldn't find"_; the header would not be present in situations `b.` and `c.`, so the client can fall back to doing a `GET` request in that case.

I should note that current versions of the registry (at least https://github.com/docker/distribution) does not send the header on `HEAD` requests if the manifest is not found, so the client would have to perform two requests for those;

```console
$ curl -X HEAD  -I \
   -H 'Accept: application/vnd.docker.distribution.manifest.v2+json' \
   -H 'Accept: application/vnd.docker.distribution.manifest.list.v2+json' \
   -H 'Accept: application/vnd.docker.distribution.manifest.v1+json' \
   http://localhost:5000/v2/nosuchrepo/manifests/latest

HTTP/1.1 404 Not Found
Content-Type: application/json
Docker-Distribution-Api-Version: registry/2.0
X-Content-Type-Options: nosniff
Date: Wed, 11 Nov 2020 21:12:11 GMT
Content-Length: 96
```

Reply:
@thaJeztah For now, we're requiring that HEAD is supported (MUST), but if a manifest is not found the check for the `Docker-Content-Digest` header is not performed.

Reply:
@thaJeztah For now, we're requiring that HEAD is supported (MUST), but if a manifest is not found the check for the `Docker-Content-Digest` header is not performed.

Reply:
>The situation is that any current registry may not support HEAD requests, and because of that return a 404 Not found.

Technically, you should return 405 (Method not Allowed) instead of 404 if you don't support HEAD requests (or any other method for that matter). I know this ideal, but we could make that the differentiator for explicitly not supporting HEAD.